wordpress-mobile/gutenberg-mobile
# Fix crash and side effects on double rotation with modal

Fixes https://github.com/wordpress-mobile/WordPress-Android/issues/10227

## Context
This PR borrows ideas from this older PR from @mkevins: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1288 

## Related PRs
Gutenberg: https://github.com/WordPress/gutenberg/pull/20860
WordPress-Android: wordpress-mobile/WordPress-Android#10383

## Description
The changes is this PR can be broken into 

- **Part 1:** simple fix for https://github.com/wordpress-mobile/WordPress-Android/issues/10227 and 
- **Part 2:** a fix for a side effect of **Part 1** that was causing modals to not reopen after rotate. 

**Part 1:** Identical to https://github.com/wordpress-mobile/gutenberg-mobile/pull/1288 (see the Description section there). Basically since we are creating a new Activity on orientation change we need to make a call to `ReactInstanceManager.onHostDestroy()`which will perform any required cleanup by our `ReactContext's` `ActivityEventListeners`. This includes removing references from our Modal to the old Activity, which fixes https://github.com/wordpress-mobile/WordPress-Android/issues/10227.

**Part 2:** of this PR is to fix some issues where modals would no longer respond after we performed the required cleanup mentioned above (see discussion here: https://github.com/wordpress-mobile/WordPress-Android/pull/10383#issuecomment-524366196). The simplest way I found to do this was to emit a message to our React Native Bridge that we are closing any open modals. Our React Native `BottomSheet` component that contains the `react-native-modal` will listen for these events on Android and perform a call to its `onClose` method which notifies the ReactNative state that the modal is ready to be opened again (see Gutenberg PR: https://github.com/WordPress/gutenberg/pull/20860). 

Alternatively, we could test an update to the [ReactModalHostView](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/modal/ReactModalHostView.java) class in the facebook `react-native` repo that adds a call to the [mOnRequestCloseListener](https://github.com/facebook/react-native/blob/10c39847a70545806dca3e1c639a46be05365332/ReactAndroid/src/main/java/com/facebook/react/views/modal/ReactModalHostView.java#L79) before Dialog cleanup occurs in its `onHostDestroy` method. I have not tested this yet, but I think proposing the change to Open Source library would be complicated because the way we utilize our React Native instance inside a "headless" fragment is non standard, so this use case may not apply to most people. 

## To test:
Tested in WordPress Android app, as well as in the Android/iOS example apps. See steps and screenshot below.

Testing steps (same apply for opening the inserter Modal, or clicking insert Image/Video to Open a modal)

1. Add an Image block and add an image from device or media library.
2. While the image block is selected, press the "gear" button to open up the Image Settings.
3. Rotate the Device.
4. Click the gear button again, the Modal should reopen as normal.
5. Rotate the Device again => the app should behave normally without crashing.

## Screenshot:
![not-crashing-rotate-modal-android](https://user-images.githubusercontent.com/1103838/76672052-650afb00-6557-11ea-9c72-31d374b6db3c.gif)

PR submission checklist:
